### ISSUE TITLE
Does vkGetDeviceQueue count as a VkQueue read access

### ISSUE THREAD

**Issue by krOoze** [KhronosGroup/Vulkan-Docs#1254]
From https://stackoverflow.com/questions/61310092/vulkan-theaded-application-get-error-message-on-queue-submissions-under-mutex

Does `vkGetDeviceQueue` count as a `VkQueue` read access? I.e. can `vkGetDeviceQueue` be executed with `vkQueueSubmit` in parallel for the same queue?

Also, it is not immediatelly clear from the spec that it will return the identical `VkQueue` handles from subsequent calls to `vkGetDeviceQueue`.

**Comment by Tobski** [comment]
> Does `vkGetDeviceQueue` count as a `VkQueue` read access? I.e. can `vkGetDeviceQueue` be executed with `vkQueueSubmit` in parallel for the same queue?

No, that seems like a validation issue - it's reading from the Device, to obtain a queue handle.

> Also, it is not immediatelly clear from the spec that it will return the identical `VkQueue` handles from subsequent calls to `vkGetDeviceQueue`.

As per https://www.khronos.org/registry/vulkan/specs/1.2/html/vkspec.html#fundamentals-commandsyntax-results-lifetime:

> Information is retrieved from the implementation with commands of the form vkGet* and vkEnumerate*.
>
> Unless otherwise specified for an individual command, the results are invariant; that is, they will remain unchanged when retrieved again by calling the same command with the same parameters, so long as those parameters themselves all remain valid.

So yes, they must be identical.

**Comment by krOoze** [comment]
Right, there's no validation issue; the bug was of somewhat different category in layers. They already assume it does not count as read.

> If two commands operate on the same object

Here it feels it is not specified enough in which cases "command operate on an object". By what should I judge whether `vkGetDeviceQueue` "operates" on the `VkQueue` or not?

> So yes, they must be identical.

Right, right.

Another issue that was raised is whether it is guaranteed each `VkQueue` is unique for each unique set of parameters. It is also slightly non-obvious because it is not usual `vkCreate*` command. I think it is safe to assume on `VkDeviceQueueCreateInfo` equals one `VkQueue` (because it is dispatchable handle).

**Comment by Tobski** [comment]
> Here it feels it is not specified enough in which cases "command operate on an object". By what should I judge whether vkGetDeviceQueue "operates" on the VkQueue or not?

I tried to see if there was anything ambiguous here and thus if anything could be clarified, and I really couldn't find anything that wasn't already clear; vkGetDeviceQueue is pretty explicit that all it does is retrieve a handle. Is there anything specific that you think is ambiguous in the spec that could be clarified?

> Another issue that was raised is whether it is guaranteed each VkQueue is unique for each unique set of parameters. It is also slightly non-obvious because it is not usual vkCreate* command. I think it is safe to assume on VkDeviceQueueCreateInfo equals one VkQueue (because it is dispatchable handle).

The text in the fundamentals chapter (which admittedly may have changed since you filed this) states:

> ...each object of a non-dispatchable type created on that device must have a handle value that is unique among objects created on that device.

There's notably no reference to specific vkCreate* commands there, and we do state in the spec that you create queues with the device; so this is already specified. I don't think it's any less of an inconvenient piece of text to find than it is for a create command either. (i.e. it could be better but that seems orthogonal)
